### PR TITLE
#16: introduced IJobExecutionContext and IJobExecutionContextFactory

### LINKED ISSUE #16: All Jobs instances are resolved from single service provider scope in asp.net core
According to https://github.com/TinkoffCreditSystems/Horarium/blob/master/src/Horarium.AspNetCore/JobFactory.cs#L16-L24 and https://github.com/TinkoffCreditSystems/Horarium/blob/master/src/Horarium/Handlers/ExecutorJob.cs#L52-L56 all jobs are resolved from captured `IServiceProvider` in `JobsFactory` constructor. This is because scope created at https://github.com/TinkoffCreditSystems/Horarium/blob/master/src/Horarium.AspNetCore/JobFactory.cs#L23 never used to resolve jobs, as its return type is `IDisposable`. Instead, jobs should be resolved from created scope like this:
`_serviceProvider.CreateScope().ServiceProvider.GetService(type)`
In particular, it affects using Horarium with EFCore, because DbContext should not to be used by multiple threads at one time.

### PR DESCRIPTION
**Pull request by wite27**
See #16 
@Bobreshovr take a look on my proposal of fix #16, please. If it's ok i'll fix up tests :)

### CODE REVIEW

**Comment by Bobreshovr** [comment]
Will this realisation work for other DI? Example `SimpleInjector`

File: src/Horarium.AspNetCore/RegistrationHorariumExtension.cs [hunk 1]
```diff
@@ -60,9 +60,9 @@ public static void StartHorariumServer(this IServiceProvider serviceProvider)
 
         private static void PrepareSettings(HorariumSettings settings, IServiceProvider serviceProvider)
         {
-            if (settings.JobFactory is DefaultJobFactory)
+            if (settings.JobExecutionContextFactory is DefaultJobExecutionContextFactory)
```
Comment (Bobreshovr) [inline src/Horarium.AspNetCore/RegistrationHorariumExtension.cs]: Name of settings is very difficult :(

Comment (wite27) [reply src/Horarium.AspNetCore/RegistrationHorariumExtension.cs]: What about `IJobScopeFactory` (and `IJobScope`)?

Comment (Bobreshovr) [reply src/Horarium.AspNetCore/RegistrationHorariumExtension.cs]: IJobScopeFactory is better 👍

**Comment by Bobreshovr** [comment]
Could you please add documentation about this factory? With example `SimpleInjector`. It would be nice.

**Comment by wite27** [comment]
> Will this realization work for other DI? Example `SimpleInjector`

I suppose, yes, but I'll take a look anyway.
> Could you please add documentation about this factory? With example SimpleInjector. It would be nice.

Of cource. Any plans to add new assembly `Horarium.SimpleInjector`?

**Comment by Bobreshovr** [comment]
> Of cource. Any plans to add new assembly `Horarium.SimpleInjector`?

Not now, but if someone wants to add `SimpleInjector`. It should be easy.

**Pull request merged**